sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,550 @@
+ */
+typedef enum {
+  /**
+   * Rule Locking Bypass disabled (mseccfg.RLB = 0).
+   */
+  kEpmpRlbToggleDisabled,
+
+  /**
+   * Rule Locking Bypass enabled (mseccfg.RLB = 1).
+   */
+  kEpmpRlbToggleEnabled,
+} epmp_rlb_toggle_t;
+
+/**
+ * Enable or disable the Rule Locking Bypass (RLB) feature.
+ *
+ * When enabled (mseccfg.RLB = 1) the Rule Locking Bypass features allows
+ * locked entries to be modified. If any PMP entries are locked and RLB
+ * is disabled (mseccfg.RLB = 0) then it is no longer possible to enable
+ * RLB. RLB will be set as requested or an error will be returned.
+ *
+ * @param toggle Whether RLB should be enabled or disabled.
+ * @return `epmp_result_t`
+ */
+epmp_result_t epmp_set_rule_locking_bypass(epmp_rlb_toggle_t toggle);
+
+/**
+ * Get the status of the Rule Locking Bypass (RLB) feature.
+ *
+ * When enabled (mseccfg.RLB = 1) the Rule Locking Bypass features allows
+ * locked entries to be modified. If any PMP entries are locked and RLB
+ * is disabled (mseccfg.RLB = 0) then it is no longer possible to enable
+ * RLB. RLB will be set as requested or an error will be returned.
+ *
+ * @param[out] toggle Whether RLB is enabled or disabled.
+ * @return `epmp_result_t`
+ */
+epmp_result_t epmp_get_rule_locking_bypass(epmp_rlb_toggle_t *toggle);
+
+/**
+ * A copy of EPMP control register state.
+ */
+typedef struct epmp_state {
+  /**
+   * PMP configuration values (pmp0cfg - pmp15cfg).
+   *
+   * These configuration values are stored in registers pmpcfg0 - pmpcfg3.
+   *
+   * Each 8-bit configuration value is encoded as follows:
+   *
+   * Layout:
+   *
+   *   +---+-------+-------+---+---+---+
+   *   | L |   0   |   A   | X | W | R |
+   *   +---+-------+-------+---+---+---+
+   *   8   7   6   5   4   3   2   1   0
+   *
+   * Key:
+   *
+   *   L = Locked
+   *   A = Address-matching Mode (OFF=0, TOR=1, NA4=2, NAPOT=3)
+   *   X = Executable
+   *   W = Writeable
+   *   R = Readable
+   *
+   * Note: the interpretation of these configuration bits depends on
+   * whether Machine Mode Lockdown (mseccfg.MML) is enabled or not.
+   * See the PMP Enhancements specification for more details.
+   */
+  uint8_t pmpcfg[kEpmpNumRegions];
+
+  /**
+   * PMP address registers (pmpaddr0 - pmpaddr15).
+   *
+   * The way that address register values are interpreted differs
+   * depending on the address-matching mode (A) in the relevant pmpcfg
+   * register(s).
+   */
+  uintptr_t pmpaddr[kEpmpNumRegions];
+
+  /**
+   * Machine Security Configuration register (mseccfg).
+   *
+   *   +---...---+------+------+------+
+   *   |    0    |  RLB | MMWP |  MML |
+   *   +---...---+------+------+------+
+   *  63         3      2      1     0
+   *
+   *  Key:
+   *
+   *    RLB  = Rule Locking Bypass
+   *    MMWP = Machine Mode Whitelist Policy
+   *    MML  = Machine Mode Lockdown
+   *
+   * See the PMP Enhancements specification for more details.
+   */
+  uint64_t mseccfg;
+
+  // TODO: checksum and/or magic values?
```

Comment:
I think the checksum should not be here (so that clients don't rely on this) but the actual API.